### DISCUSSION TITLE
failed to run custom build command for `pyo3 v0.15.1`

### DISCUSSION THREAD

**Discussion by Qyanjia** [PyO3/pyo3#2073] [Questions]
I got the problem as follows:
**the configured Python interpreter version (3.5) is lower than PyO3's minimum supported version (3.6)**

but my Python version is already 3.7

![20211222141325](https://user-images.githubusercontent.com/61483129/147044754-16e37f8a-03a3-46b0-b9e1-23954a5f666e.png)

![20211222141215](https://user-images.githubusercontent.com/61483129/147044772-5fef2779-8c3a-4125-b957-a14af202b1ef.png)

![Ubuntu](https://user-images.githubusercontent.com/61483129/147044808-959d4ab2-51aa-45b1-a216-1b4ec3321ec4.png)

what should i do

**Comment by messense** [comment]
Run your build with `PYO3_PRINT_CONFIG=1` and paste the log here in text. (Please don't just paste a screenshot.)

See https://pyo3.rs/v0.15.1/building_and_distribution.html#configuring-the-python-version

**Comment by Qyanjia** [comment]
> Run your build with and paste the log here in text. (Please don't just paste a screenshot.)`PYO3_PRINT_CONFIG=1`
> 
> See https://pyo3.rs/v0.15.1/building_and_distribution.html#configuring-the-python-version

   Compiling pyo3 v0.15.1
error: failed to run custom build command for `pyo3 v0.15.1`

Caused by:
  process didn't exit successfully: `/home/qinyanjia/my_project/python/target/debug/build/pyo3-39ac1bd7d210807f/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=PYO3_CROSS
  cargo:rerun-if-env-changed=PYO3_CROSS_LIB_DIR
  cargo:rerun-if-env-changed=PYO3_CROSS_PYTHON_VERSION
  cargo:rerun-if-env-changed=PYO3_PRINT_CONFIG

  -- PYO3_PRINT_CONFIG=1 is set, printing configuration and halting compile --
  implementation=CPython
  version=3.7
  shared=true
  abi3=false
  lib_name=python3.7m
  lib_dir=/usr/local/lib
  executable=/usr/bin/python
  pointer_width=64
  build_flags=WITH_THREAD
  suppress_build_script_link_lines=false

**i run the build and get the above result , it looks like no problem. but i cant run the example that in the section (Using Python from Rust)**

error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.126auf7iuzmzhqk1.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.15bqgr347tv4ugn4.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.1a7ail26uglsjrhr.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.1qjsvrstn7wp0t6b.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.252pjhe4bksrpr88.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.28t1o3zlao9tjyd5.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.2cnupz43lc8751nq.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.2fg4f64mue2z9lny.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.2sth52qzpv8eutgs.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.30r40g8ci2zsqx6q.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.331krtk13sa5fwjl.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.34bsc49o0vgt2nxo.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.36hxkv2vw7zuaaf2.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.3a83rzsuuneiojdc.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.3etv7z4v0dt5lwp7.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.3f0ewzr30jwt93xp.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.3nqq8htj9eeal1hl.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.3wtflj1lv9meaotm.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4and4rof9wk05aka.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4dpkpa0l1n10h7sb.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4g84zmi0s22nt8do.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4pqe3c2o5l65j8i2.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4wzz8s1c87oq2qg9.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4x0smhazg5lm99tq.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.4y43s2e7e6gp773b.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.50mfxf5zzeefw5s5.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.517t8h5txuygwyqb.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.58znxk2cavdnz2s2.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.59652q8fybpdgywg.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.5em29w8xjpc9e2kt.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.79914tg2omaaz6s.rcgu.o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157.5bzcnjnlkh57ds3z.rcgu.o" "-Wl,--as-needed" "-L" "/home/qinyanjia/my_project/python/target/debug/deps" "-L" "/usr/local/lib" "-L" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/qinyanjia/my_project/python/target/debug/deps/libpyo3-710d816125d38ca3.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libparking_lot-e676e2ed76fc2dcb.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libparking_lot_core-4c59f6448432095c.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libsmallvec-625b889d3db5168f.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/liblock_api-a82c14b263ee1440.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libscopeguard-e4ce1e91fe43df34.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libinstant-24f006b0a019c4f1.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libcfg_if-0c646b04edbea3dc.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/liblibc-f9a51910ecabca80.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libunindent-f93ed6820ddb0f2f.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libpaste-9fab1314b1b91fac.rlib" "/home/qinyanjia/my_project/python/target/debug/deps/libindoc-eb33c75adf52dc51.rlib" "-Wl,--start-group" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-65a28bf1738424c0.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-ba99c63b775f734c.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2c795be2d1f43514.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-4225fa5bc39268de.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-a6e9f64eae6e45dd.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-b1c963add6e8b853.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-541ca82647a7fb7c.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-b93187d0f71da706.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-12352e7b2d1a318e.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-3787a903f1bee791.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-fdd9bd7e7b7760b0.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-4274f5e0121314fe.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-496757222b2a1fa7.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-e68feee69f5a82d8.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-5d6d15fab36665ef.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-10916d3506f78066.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-f6aebf5fbe06d09a.rlib" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-4e1b16fc9c111ada.rlib" "-Wl,--end-group" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-02203e01b7df4fdd.rlib" "-Wl,-Bdynamic" "-lpython3.7m" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/home/qinyanjia/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/qinyanjia/my_project/python/target/debug/deps/python-3af51c414efc7157" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs"
  = note: /usr/bin/ld: /usr/local/lib/libpython3.7m.a(abstract.o): relocation R_X86_64_32S against `_Py_NotImplementedStruct' can not be used when making a shared object; recompile with -fPIC
          /usr/local/lib/libpython3.7m.a: error adding symbols: Bad value
          collect2: error: ld returned 1 exit status
          

error: could not compile `python` due to previous error

**Comment by messense** [comment]
Did you enable the `extension-module` feature of pyo3? How did you install your Python interpreter? Did it built with `--enable-shared`?

**Comment by Qyanjia** [comment]
> 

wow ,i will check it right now!  I didnot build with --enable-shared

**Comment by messense** [comment]
Converting this issue to discussion since it doesn't look like a pyo3 bug.

**Reply by Qyanjia** [reply]
I solved this problem , thx your patience . by the way,  lao ge niu bi ！！！
